02-client-refactor added support for multiple consensus states inside a header

nadeemb53 opened the pull request:
A beefy or grandpa header submitted on-chain can potentially contain headers with multiple consensus states. This PR refactors the client interface to support beefy and grandpa light clients.

bluele commented:
Thanks for your PR!

My understanding is that in ICS-02 and ibc-go, each client implementation currently has the responsibility for client state updates. (ref. https://github.com/cosmos/ibc-go/pull/1871) I would like to apply the same design to ibc-solidity as much as possible. Would this design change help with beefy and grandpa client implementations? Also, if you have already considered it, please tell us why it was difficult.

(however, there may be some things I missed about Gas cost or EVM constraints)
cc: @siburu @3100

siburu commented:
@bluele 
> each client implementation currently has the responsibility for client state updates

Do you mean that `IClient::updateClient` itself should call `host.setClientState`?

siburu commented:
Aside from that, I think you should move the definition of the handwritten `ConsensusUpdate.Result` to a different location.
The `contracts/types` directory contains only `.sol` files generated from `.proto` files.

@bluele Where is an appropriate place to move it to?

Wizdave97 commented on contracts/core/IBCClient.sol:
```diff
@@ -26,43 +27,89 @@ library IBCClient {
+            .updateState(host, msg_.clientId, clientStateBytes, msg_.header);
+
         //// persist states ////
         host.setClientState(msg_.clientId, clientStateBytes);
```
@nadeemb53  we can still work with @bluele 's suggestion, instead of having the update_state method return a `ConsensusUpdateResult`,
let's do all the state updates inside `update_state` , so that way each light client can store the consensus state however they want.

So you just have to move the calls to `setClientState`, `setConsensusState`, `setProcessedTime`, `setProcessedHeight`, into `update_state` for the beefy

Wizdave97 replied:
For now add some documentation to the `update_state` function that it should do all the state persisting internally and return an array of consensus heights.

nadeemb53 replied:
Addressed this in the recent commit

nadeemb53 commented on contracts/core/IClient.sol:
```diff
@@ -15,17 +14,29 @@ interface IClient {
+            bytes memory newClientStateBytes,
+            ConsensusUpdate.Result[] memory results
+        );
 
```
Addresses this in the recent commit

nadeemb53 commented:
Hey @bluele, thanks for your comment. I have implemented it according to your suggestion.

@siburu thanks for your comment, I have removed `consensusUpdate.Result` as it was not needed anymore.

bluele commented:
hey @nadeemb53, I realized there is one problem with moving the state updates to each client implementation. 

The set* functions of the host cannot be called from Light Client contracts due to `onlyIBCModule` modifier. To prevent this, we need to use delegatecall for the Light Client contract call.

This have already been fixed in PR #116 . Therefore, please rebase the branch with the latest main. Also, could you minimize the changes in the commits? (e.g. the changes regarding code format are not needed in this PR)

nadeemb53 commented:
@bluele thanks for your changes. I've rebased and followed it to further fix my PR.

bluele commented on contracts/core/IClient.sol:
```diff
@@ -20,12 +20,33 @@ interface IClient {
     ) external view returns (Height.Data memory, bool);
 
-    function checkHeaderAndUpdateState(
+    function verifyClientMessage(
```
I understand that the interface functions(verifyClientMessage,updateState, checkForMisbehaviour, updateStateOnMisbehaviour) come from ibc-go, but I wonder if they should be added in ibc-solidity.

Since there is no abstract type like go's interface in solidity, client state needs to be represented as serialized bytes, except in the corresponding Light Client contract. Thus, this splitting of update_client function causes deserialization in each function. The costs are often high, so they are not negligible in solidity. So it is not necessary to apply such abstraction to the IClient.

Therefore, I think it would be better to simply rename `checkHeaderAndUpdateState` function as `updateClient` and each client implementation should handle misbehaviour in the function. Please let me know if I am missing something.

nadeemb53 replied:
hey yeah, I think your point is valid. I have renamed `checkHeaderAndUpdateState` to `verifyClientMessageAndUpdateState` and rolled back the corresponding changes.

bluele commented:
Hi @nadeemb53, I've left some comments. Also, run CI and it looks like integration-test is failed.

bluele commented on contracts/core/IBFT2Client.sol:
```diff
@@ -100,20 +100,87 @@ contract IBFT2Client is IClient {
+        );
+        host.setProcessedTime(clientId, parsedHeader.height, block.timestamp);
+        host.setProcessedHeight(clientId, parsedHeader.height, block.timestamp);
+        heights[0] = parsedHeader.height;
```
@nadeemb53 can you fix this as the followings?
```
heights = new Height.Data[](1);
heights[0] = parsedHeader.height;
```

nadeemb53 commented:
Hi @bluele , could you please run the CI?

bluele commented:
@nadeemb53 done. Please squash your commits into one commit.

nadeemb53 commented:
done @bluele

bluele commented:
@nadeemb53 Great work! I've merged this.
